pkg/streamingpromql/operators/aggregations/topkbottomk/range_query.go

```diff
@@ -0,0 +1,505 @@
+
+	g := t.remainingInnerSeriesToGroup[0]
+	t.remainingInnerSeriesToGroup = t.remainingInnerSeriesToGroup[1:]
+	if err := t.accumulateIntoGroup(nextSeries, g); err != nil {
+		return err
+	}
+
+	types.FPointSlicePool.Put(nextSeries.Floats, t.MemoryConsumptionTracker)
+
+	return nil
+}
+
+func (t *RangeQuery) emitIgnoredHistogramsAnnotation() {
+	if t.haveEmittedIgnoredHistogramsAnnotation {
+		return
+	}
+
+	if t.IsTopK {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("topk", t.expressionPosition))
+	} else {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("bottomk", t.expressionPosition))
+	}
+
+	t.haveEmittedIgnoredHistogramsAnnotation = true
+}
+
+func (t *RangeQuery) accumulateIntoGroup(data types.InstantVectorSeriesData, g *rangeQueryGroup) error {
+	groupSeriesIndex := g.seriesRead()
+
+	if g.seriesForTimestamps == nil {
+		var err error
+		g.seriesForTimestamps, err = intSliceSlicePool.Get(t.TimeRange.StepCount, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.seriesForTimestamps = g.seriesForTimestamps[:t.TimeRange.StepCount]
+	}
+
+	if g.series == nil {
+		var err error
+		g.series, err = rangeQuerySeriesSlicePool.Get(g.totalSeries, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+	}
+
+	g.series = g.series[:groupSeriesIndex+1]
+	thisSeries := &g.series[groupSeriesIndex]
+
+	for _, p := range data.Floats {
+		timestampIndex := t.TimeRange.PointIndex(p.T)
+		limit := t.k[timestampIndex]
+
+		if limit == 0 {
+			continue
+		}
+
+		if g.seriesForTimestamps[timestampIndex] == nil {
+			// This is the first time we've seen a point for this timestamp, create the list of source series.
+			maximumPossibleSeries := min(limit, int64(g.totalSeries))
+
+			var err error
+			g.seriesForTimestamps[timestampIndex], err = types.IntSlicePool.Get(int(maximumPossibleSeries), t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+		}
+
+		populateSeriesSlices := func() error {
+			lastPointIndex := t.TimeRange.PointIndex(data.Floats[len(data.Floats)-1].T)
+			sliceLength := int(lastPointIndex + 1)
+
+			var err error
+			thisSeries.shouldReturnPoint, err = types.BoolSlicePool.Get(sliceLength, t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+
+			thisSeries.shouldReturnPoint = thisSeries.shouldReturnPoint[:sliceLength]
+
+			thisSeries.values, err = types.Float64SlicePool.Get(sliceLength, t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+
+			thisSeries.values = thisSeries.values[:sliceLength]
+
+			return nil
+		}
+
+		recordUsedPoint := func() error {
+			thisSeries.pointCount++
+
+			if thisSeries.shouldReturnPoint == nil {
+				if err := populateSeriesSlices(); err != nil {
+					return err
+				}
+			}
+
+			thisSeries.shouldReturnPoint[timestampIndex] = true
+			thisSeries.values[timestampIndex] = p.F
+
+			return nil
+		}
+
+		if len(g.seriesForTimestamps[timestampIndex]) == int(limit) {
```

Comment:
(nit) point selection/interacting with the heap could be split out into own method

Reply:
I've moved the other methods out in 54bba6e, which should help make this a little shorter - how's this now?

Reply:
I still feel this if/else block could be extracted so we have `accumulateIntoGroup` dealing with getting the series for the group, and a new method such as `updateHeapForGroup` or similar

Reply:
Done in https://github.com/grafana/mimir/pull/10679